### ISSUE TITLE
Files compacted by size rather than row count

### ISSUE THREAD

**Issue by jarohen** [xtdb/crux#3845]
We originally assumed (for simplicity) that picking a reasonable number of rows for a file would be a good-enough proxy for file size - we knew it wasn't going to be accurate, and that row-sizes would vary significantly, but this now doesn't look viable in practice.

* We will consider a file 'full' at around 64MB-128MB, to be tuned.
* All three major cloud providers include size in their list-objects calls, so we no longer need to retain this in the file name. We will want to cache the entire file-tree (file names and sizes) on each node by querying from listObjects on startup and then following the file-list notification topic (which already contains the file sizes). 
  * Separately, we can potentially also cache which files in the tree are live at this point, to save us re-calculating this on scan. (raise a separate card if this isn't straightforward to include as part of this one)
* With this card, we'd also like to remove the first-row/next-row in the file-names. These were originally a proxy for the transaction id - historically, it was because one of the original indexes required each row to have an id - but we should now be able to use tx-id and remove the complexity.
* We also don't need the first row - the next-tx-id and the fact that a file is full/not-full should be sufficient to determine which files are live.

L0/L1: `{l00,l01}/<next-tx-id>.arrow`
L2+: `l02/<iidpart>/<next-tx-id>.arrow`

In #4047, we expect this to change slightly:

L0/L1: `{l00,l01}/<next-tx-id>.arrow`
L2: `l02/<timepart>/<next-tx-id>.arrow`
L3+: `l03/<timepart/<iidpart>/<next-tx-id.arrow>`

We currently compact a non-full L1 file and N * L0 files together if there are more than one uncompacted L0 file. With row-counts, the row-count of the output file was guaranteed to be the sum of the row-counts of the inputs files. When compacting by size, though, we can no longer do this - the size of the output file may be smaller than the sum of its parts (it won't be bigger), because the schema won't be fully duplicated.

To preserve determinism without co-ordination between different nodes in the cluster, we can no longer simply compact all of the uncompacted L0 files (because we no longer know the output size before doing the merge) but we'd still like to merge as many as we can. We will therefore include as many L0 files as we can without going over the size limit first. Then, if we're still under the limit, we upload that file anyway (as a non-full file), and then incorporate L0 files one at a time.

**Comment by refset** [comment]
Related/dupe https://github.com/xtdb/xtdb/issues/2816

**Comment by jarohen** [comment]
@refset yep, I recall that one - I think it's slightly separate as this one only affects the compactor, I think Dan was more after making the indexer byte-related rather than row-related (not something I was planning to tackle here).

Definitely related though :slightly_smiling_face:
